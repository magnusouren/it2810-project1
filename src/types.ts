export type CategoryType =
  | 'Ordinary drink'
  | 'Cocktail'
  | 'Shake'
  | 'Other / Unknown'
  | 'Cocoa'
  | 'Shot'
  | 'Coffee / Tea'
  | 'Homemade Liqueur'
  | 'Punch / Party Drink'
  | 'Beer'
  | 'Soft Drink';

export type Alcoholic = 'Alcoholic' | 'Non alcoholic' | 'Optional alcohol';

export interface Ingredient {
  ingredient: string;
  measure: string;
}

export interface Drink {
  drinkId: string;
  strDrink: string;
  ingredients: Array<Ingredient>;
  strGlass: string;
  strInstructions: string;
  strDrinkThumb: string;
  strCategory: string;
  strAlcoholic: Alcoholic;
}
