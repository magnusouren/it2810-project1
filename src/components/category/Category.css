.category {
  flex-grow: 1;
  margin: 0 auto;
  padding: 3rem 1rem 1rem;
  background-color: #fff;
  border: solid 1px #eaeaea;
  border-radius: 8px;
  text-decoration: none #f2c2a6;
  transition: 0.2s ease-in-out;
}

.category h3 {
  color: #000;
  font-family: Lato, sans-serif;
  font-weight: 500;
}

.category:hover {
  border: solid 1px #000;
  cursor: pointer;
  text-decoration: underline #f2c2a6;
}

@media screen and (width <= 460px) {
  .category {
    width: 100%;
  }
}
