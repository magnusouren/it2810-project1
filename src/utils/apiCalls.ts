import axios from 'axios';

export const fetchCocktailById = async (id: string) => {
  const response = await axios.get(`https://www.thecocktaildb.com/api/json/v1/1/lookup.php?i=${id}`);
  return response.data;
};
