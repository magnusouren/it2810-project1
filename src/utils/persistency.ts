import { DrinkOfTheDay } from '../types';

const addFavorite = (id: string) => {
  const favorites = getFavorites();
  favorites.push(id);
  localStorage.setItem('drinkFavorites', JSON.stringify(favorites));
};

export const getFavorites = () => {
  const favorites = localStorage.getItem('drinkFavorites');
  if (favorites) {
    return JSON.parse(favorites);
  }
  return [];
};

const removeFavorite = (id: string) => {
  const favorites = getFavorites();
  const newFavorites = favorites.filter((favorite: string) => favorite !== id);
  localStorage.setItem('drinkFavorites', JSON.stringify(newFavorites));
};

export const toggleFavorite = (id: string) => {
  const favorites = getFavorites();
  if (favorites.includes(id)) {
    removeFavorite(id);
  } else {
    addFavorite(id);
  }
};

export const isFavorite = (id: string) => {
  const favorites = getFavorites();
  return favorites.includes(id);
};

export const getDrinkOfTheDay = (date: string) => {
  const drinkOfTheDay = localStorage.getItem(date);
  if (drinkOfTheDay) {
    return JSON.parse(drinkOfTheDay);
  }
  return null;
};

export const setDrinkOfTheDay = (date: string, drink: DrinkOfTheDay) => {
  localStorage.setItem(date, JSON.stringify(drink));
};
