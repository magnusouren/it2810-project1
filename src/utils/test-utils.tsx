import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { render } from '@testing-library/react';
import { BrowserRouter, MemoryRouter, Route, Routes } from 'react-router-dom';

import { DrinkCard } from '../components/drink/DrinkCard';

// Function to render a component wrapped inside a browserrouter for testing purposes
export const renderWithRouter = (component: JSX.Element) => render(<BrowserRouter>{component}</BrowserRouter>);

const queryClient = new QueryClient();

export const renderWithRouterAndQueryClient = (component: JSX.Element) =>
  render(
    <QueryClientProvider client={queryClient}>
      <BrowserRouter>{component}</BrowserRouter>
    </QueryClientProvider>,
  );

export const renderWithRouterQueryClientAndDrinkId = (id: string) =>
  render(
    <QueryClientProvider client={new QueryClient()}>
      <MemoryRouter initialEntries={[`/drink/${id}`]}>
        <Routes>
          <Route path='/drink'>
            <Route path=':id' element={<DrinkCard />} />
          </Route>
        </Routes>
      </MemoryRouter>
    </QueryClientProvider>,
  );
