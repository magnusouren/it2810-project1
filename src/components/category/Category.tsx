import './Category.css';

import { FC } from 'react';
import { Link } from 'react-router-dom';

import { CategoryType } from '../../types';

interface CategoryProps {
  category: CategoryType;
}

export const Category: FC<CategoryProps> = ({ category }) => (
  <Link to={`/search/`} state={category} className='category'>
    <h3>{category}</h3>
  </Link>
);
