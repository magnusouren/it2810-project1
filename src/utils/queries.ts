import axios, { isCancel } from 'axios';

import { Drink, DrinkOfTheDay, Ingredient } from '../types';
import { setDrinkOfTheDay } from './persistency';

export const fetchDrinkById = async (id?: string) => {
  if (!id) {
    return null;
  }
  const response = await axios
    .get(`https://www.thecocktaildb.com/api/json/v1/1/lookup.php?i=${id}`)
    .then((response) => {
      // Extract drink data
      const drinkData = response.data.drinks[0];
      const ingredients: Array<Ingredient> = [];
      const drink: Drink = {
        idDrink: drinkData.idDrink,
        strDrink: drinkData.strDrink,
        strGlass: drinkData.strGlass,
        ingredients: [],
        strInstructions: drinkData.strInstructions,
        strDrinkThumb: drinkData.strDrinkThumb,
        strCategory: drinkData.strCategory,
        strAlcoholic: drinkData.strAlcoholic,
      };

      // Extract ingredients
      for (let i = 1; i <= 15; i++) {
        const ingredient: string = drinkData[`strIngredient${i}`];
        const measure: string = drinkData[`strMeasure${i}`];
        if (ingredient) {
          ingredients.push({ ingredient, measure });
        } else {
          break;
        }
      }

      drink.ingredients = ingredients;
      return drink;
    })
    .catch((error) => {
      if (isCancel(error)) {
        console.log('Request cancelled', error.message);
        return null;
      }
      if (error instanceof TypeError) {
        return null;
      } else {
        console.error('Error fetching drink:', error);
        return null;
      }
    });
  return response;
};

export const fetchDrinkOfTheDay = async (currentDate: string) => {
  return axios
    .get('https://www.thecocktaildb.com/api/json/v1/1/random.php')
    .then((response) => {
      const randomDrink = response.data.drinks[0];

      const drink: DrinkOfTheDay = {
        drinkId: randomDrink.idDrink,
        strDrink: randomDrink.strDrink,
        strDrinkThumb: randomDrink.strDrinkThumb,
        strCategory: randomDrink.strCategory,
        strGlass: randomDrink.strGlass,
        strAlcoholic: randomDrink.strAlcoholic,
      };

      // Store the new random drink ID in local storage with the current date
      setDrinkOfTheDay(currentDate, drink);

      return drink;
    })
    .catch((error) => {
      console.error('Error fetching drink:', error);
      return null;
    });
};
