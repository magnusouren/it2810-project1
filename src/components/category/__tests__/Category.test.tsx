import { act, screen } from '@testing-library/react';

import { CategoryType } from '../../../types';
import { renderWithRouterAndQueryClient } from '../../../utils/test-utils';
import { Category } from '../Category';

const categories: CategoryType[] = [
  'Ordinary Drink',
  'Cocktail',
  'Shake',
  'Other / Unknown',
  'Cocoa',
  'Shot',
  'Coffee / Tea',
  'Homemade Liqueur',
  'Punch / Party Drink',
  'Beer',
  'Soft Drink',
];

describe('Category', () => {
  it('Should match snapshot', () => {
    const { container } = renderWithRouterAndQueryClient(<Category category={'Ordinary Drink'} />);
    expect(container).toMatchSnapshot();
  });

  it('Should show category when it is passed as prop', () => {
    renderWithRouterAndQueryClient(<Category category={'Ordinary Drink'} />);
    expect(screen.getByText(/Ordinary Drink/i)).toBeDefined();
  });

  it.each(categories)('Should navigate correct when "%s" is clicked', async (category) => {
    renderWithRouterAndQueryClient(<Category category={category} />);
    act(() => {
      screen.getByText(category).click();
    });
    expect(window.location.pathname).toBe('/search/');
    expect(window.history.state.usr).toEqual(category);
  });

  it.each(categories)('Should use "%s" as state in navigation when clicked', async (category) => {
    renderWithRouterAndQueryClient(<Category category={category} />);
    act(() => {
      screen.getByText(category).click();
    });
    expect(window.history.state.usr).toEqual(category);
  });
});
