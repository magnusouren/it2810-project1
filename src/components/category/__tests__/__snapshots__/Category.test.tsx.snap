// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Category > Should match snapshot 1`] = `
<div>
  <a
    class="category"
    href="/search/"
  >
    <h3>
      Ordinary Drink
    </h3>
  </a>
</div>
`;
